File: skills/guidelines/examples/schema-example.py
"""
Example: SQLAlchemy model + Pydantic schema separation.

SQLAlchemy handles persistence. Pydantic handles validation at boundaries.
Keep them separate: models define what's stored, schemas define what's accepted
and returned.
"""

import enum
from datetime import datetime

from sqlalchemy import String, Enum as SAEnum, DateTime, ForeignKey, func
from sqlalchemy.orm import Mapped, mapped_column, DeclarativeBase
from pydantic import BaseModel, Field


# --- SQLAlchemy models (persistence) ---


class Base(DeclarativeBase):
    pass


class PassStatus(str, enum.Enum):
    active = "active"
    completed = "completed"
    cancelled = "cancelled"


class Pass(Base):
    __tablename__ = "passes"

    id: Mapped[str] = mapped_column(String, primary_key=True)
    student_id: Mapped[str] = mapped_column(
        String, ForeignKey("students.id"), nullable=False
    )
    teacher_id: Mapped[str] = mapped_column(
        String, ForeignKey("teachers.id"), nullable=False
    )
    status: Mapped[PassStatus] = mapped_column(
        SAEnum(PassStatus), default=PassStatus.active, nullable=False
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )


# --- Pydantic schemas (boundaries) ---


class CreatePassInput(BaseModel):
    student_id: str = Field(..., description="UUID of the student")
    teacher_id: str = Field(..., description="UUID of the requesting teacher")
    notes: str | None = Field(None, max_length=500)


class PassResponse(BaseModel):
    id: str
    student_id: str
    teacher_id: str
    status: PassStatus
    created_at: datetime

    model_config = {"from_attributes": True}

File: skills/guidelines/examples/schema-example.ts
// Example: Schema and data model pattern
// Drizzle table definitions paired with Zod schemas for validation.
// The Zod schema is the source of truth for input validation.
// Drizzle handles persistence; Zod handles boundaries.

import { pgTable, uuid, varchar, timestamp, pgEnum } from "drizzle-orm/pg-core";
import { z } from "zod";

// Drizzle enum + table definition
export const passStatusEnum = pgEnum("pass_status", [
  "active",
  "completed",
  "expired",
  "revoked",
]);

export const passes = pgTable("passes", {
  id: uuid("id").defaultRandom().primaryKey(),
  studentId: uuid("student_id").notNull(),
  teacherId: uuid("teacher_id").notNull(),
  destination: varchar("destination", { length: 255 }).notNull(),
  status: passStatusEnum("status").notNull().default("active"),
  createdAt: timestamp("created_at").defaultNow().notNull(),
  completedAt: timestamp("completed_at"),
});

// Zod schema for creating a pass (input validation at boundaries)
export const createPassSchema = z.object({
  studentId: z.string().uuid(),
  teacherId: z.string().uuid(),
  destination: z.string().min(1).max(255),
});

// Zod schema for the full pass shape (used for response typing)
export const passSchema = z.object({
  id: z.string().uuid(),
  studentId: z.string().uuid(),
  teacherId: z.string().uuid(),
  destination: z.string(),
  status: z.enum(["active", "completed", "expired", "revoked"]),
  createdAt: z.coerce.date(),
  completedAt: z.coerce.date().nullable(),
});
